Make RemoteRenderingData::cacheFont reconstruct a Font using Attributes, and a reference to FontCustomPlatformData.

mattwoodrow opened the pull request:
#### e36b84b4cc79b4f28020cd86825a51e361e479f3
<pre>
Cherry-pick c201cea073fd. &lt;rdar://106965215&gt;

    Make RemoteRenderingData::cacheFont reconstruct a Font using Attributes, and a reference to FontCustomPlatformData.
    <a href="https://bugs.webkit.org/show_bug.cgi?id=254490">https://bugs.webkit.org/show_bug.cgi?id=254490</a>
    &lt;rdar://106965215&gt;

    Reviewed by NOBODY (OOPS!).

    Make FontCustomPlatformData refcounted, and owned by the FontPlatformData instead of CreationData

    We currently only have the CreationData accessible from Font, but when serializing a web font,
    we need the FontCustomPlatformData in order to share the parsed font data.

    This just makes the FontCustomPlatformData the shared object on FontPlatformData, and we can
    still access the CreationData through it.

    * Source/WebCore/css/CSSFontFaceSource.h:
    * Source/WebCore/loader/cache/CachedFont.cpp:
    (WebCore::CachedFont::createCustomFontData):
    * Source/WebCore/loader/cache/CachedFont.h:
    * Source/WebCore/platform/graphics/FontCustomPlatformData.h:
    (WebCore::FontCustomPlatformData::FontCustomPlatformData):
    * Source/WebCore/platform/graphics/FontPlatformData.cpp:
    (WebCore::FontPlatformData::FontPlatformData):
    (WebCore::FontPlatformData::creationData const):
    * Source/WebCore/platform/graphics/FontPlatformData.h:
    (WebCore::FontPlatformData::customPlatformData const):
    (WebCore::FontPlatformData::creationData const): Deleted.
    * Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
    (WebCore::FontCache::systemFallbackForCharacters):
    * Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
    (WebCore::createDerivativeFont):
    (WebCore::Font::createFontWithoutSynthesizableFeatures const):
    (WebCore::Font::platformCreateScaledFont const):
    * Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
    (WebCore::FontPlatformData::FontPlatformData):
    * Source/WebCore/platform/graphics/mac/FontCustomPlatformDataMac.cpp:
    (WebCore::FontCustomPlatformData::fontPlatformData):
    (WebCore::createFontCustomPlatformData):
    (WebCore::FontCustomPlatformData::creationData const):
    * Source/WebCore/workers/WorkerFontLoadRequest.h:
    * Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
    (IPC::ArgumentCoder&lt;WebCore::Font&gt;::decodePlatformData):
    * Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
    * Source/WebKit/Shared/WebCoreArgumentCoders.h:

    Move POD serializable subset of Font into Attributes object

    In order to serialize a Font where the FontCustomPlatformData is already cached, we want to send
    the identifier of the data, plus the serialized remainder of the font.

    This creates an Attributes inner class for Font, containing the data needed to serialize the font,
    without the platform data.

    * Source/WebCore/platform/graphics/Font.cpp:
    (WebCore::Font::Font):
    (WebCore::Font::renderingResourceIdentifier const):
    * Source/WebCore/platform/graphics/Font.h:
    (WebCore::Font::isTextOrientationFallback const):
    (WebCore::Font::origin const):
    (WebCore::Font::isInterstitial const):
    (WebCore::Font::visibility const):
    (WebCore::Font::attributes const):
    * Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
    (IPC::ArgumentCoder&lt;WebCore::Font&gt;::encode):
    (IPC::ArgumentCoder&lt;Font&gt;::decode):
    (IPC::ArgumentCoder&lt;WebCore::Font::Attributes&gt;::encode):
    (IPC::ArgumentCoder&lt;Font::Attributes&gt;::decode):
    * Source/WebKit/Shared/WebCoreArgumentCoders.h:

    Add serialization and RemoteResourceCache support for FontCustomPlatformData

    This lets us cache FontCustomPlatformData objects independently of Font, so that we
    can share them between Font instances in the GPUP.

    * Source/WebKit/GPUProcess/graphics/QualifiedResourceHeap.h:
    (WebKit::QualifiedResourceHeap::add):
    (WebKit::QualifiedResourceHeap::getFontCustomPlatformData const):
    (WebKit::QualifiedResourceHeap::removeFontCustomPlatformData):
    (WebKit::QualifiedResourceHeap::releaseAllResources):
    (WebKit::QualifiedResourceHeap::checkInvariants const):
    * Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
    (WebKit::RemoteRenderingBackend::cacheFontCustomPlatformData):
    (WebKit::RemoteRenderingBackend::cacheFontCustomPlatformDataWithQualifiedIdentifier):
    * Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
    * Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
    * Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
    (WebKit::RemoteResourceCache::cacheFontCustomPlatformData):
    (WebKit::RemoteResourceCache::cachedFontCustomPlatformData const):
    (WebKit::RemoteResourceCache::releaseRenderingResource):
    * Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
    * Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
    (IPC::ArgumentCoder&lt;WebCore::FontCustomPlatformData&gt;::encodePlatformData):
    (IPC::ArgumentCoder&lt;WebCore::FontCustomPlatformData&gt;::decodePlatformData):
    * Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
    (IPC::ArgumentCoder&lt;WebCore::FontCustomPlatformData&gt;::encode):
    (IPC::ArgumentCoder&lt;FontCustomPlatformData&gt;::decode):
    * Source/WebKit/Shared/WebCoreArgumentCoders.h:
    * Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
    (WebKit::RemoteRenderingBackendProxy::cacheFontCustomPlatformData):
    * Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
    * Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
    (WebKit::RemoteResourceCacheProxy::recordFontCustomPlatformDataUse):
    (WebKit::RemoteResourceCacheProxy::prepareForNextRenderingUpdate):
    (WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdateForFonts):
    * Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

    Add serialization and RemoteResourceCache support for FontCustomPlatformData

    This lets us cache FontCustomPlatformData objects independently of Font, so that we
    can share them between Font instances in the GPUP.

    * Source/WebKit/GPUProcess/graphics/QualifiedResourceHeap.h:
    (WebKit::QualifiedResourceHeap::add):
    (WebKit::QualifiedResourceHeap::getFontCustomPlatformData const):
    (WebKit::QualifiedResourceHeap::removeFontCustomPlatformData):
    (WebKit::QualifiedResourceHeap::releaseAllResources):
    (WebKit::QualifiedResourceHeap::checkInvariants const):
    * Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
    (WebKit::RemoteRenderingBackend::cacheFontCustomPlatformData):
    (WebKit::RemoteRenderingBackend::cacheFontCustomPlatformDataWithQualifiedIdentifier):
    * Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
    * Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
    * Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
    (WebKit::RemoteResourceCache::cacheFontCustomPlatformData):
    (WebKit::RemoteResourceCache::cachedFontCustomPlatformData const):
    (WebKit::RemoteResourceCache::releaseRenderingResource):
    * Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
    * Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
    (IPC::ArgumentCoder&lt;WebCore::FontCustomPlatformData&gt;::encodePlatformData):
    (IPC::ArgumentCoder&lt;WebCore::FontCustomPlatformData&gt;::decodePlatformData):
    * Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
    (IPC::ArgumentCoder&lt;WebCore::FontCustomPlatformData&gt;::encode):
    (IPC::ArgumentCoder&lt;FontCustomPlatformData&gt;::decode):
    * Source/WebKit/Shared/WebCoreArgumentCoders.h:
    * Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
    (WebKit::RemoteRenderingBackendProxy::cacheFontCustomPlatformData):
    * Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
    * Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
    (WebKit::RemoteResourceCacheProxy::recordFontCustomPlatformDataUse):
    (WebKit::RemoteResourceCacheProxy::prepareForNextRenderingUpdate):
    (WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdateForFonts):
    * Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

    Add FontPlatformData::Attributes for the POD serializable subset.

    Similar to adding Font::Attributes, this adds FontPlatformData::Attributes for the serializable data,
    that can be combined with an (optional) FontCustomPlatformData in order to create a final object.

    Unfortunately this can&apos;t actually be used as the main storage within FontPlatformData, as we need
    to do some data conversion before serialization (and the equivalent conversion on deserialization
    can&apos;t be done without the FontCustomPlatformData).

    * Source/WebCore/platform/graphics/FontPlatformData.h:
    (WebCore::FontPlatformData::Attributes::Attributes):
    * Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
    (WebCore::findFontDescriptor):
    (WebCore::createCTFont):
    (WebCore::FontPlatformData::create):
    (WebCore::FontPlatformData::attributes const):
    * Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
    (IPC::ArgumentCoder&lt;WebCore::FontPlatformData::Attributes&gt;::encodePlatformData):
    (IPC::ArgumentCoder&lt;WebCore::FontPlatformData::Attributes&gt;::decodePlatformData):
    * Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
    (IPC::ArgumentCoder&lt;WebCore::FontPlatformData::Attributes&gt;::encode):
    (IPC::ArgumentCoder&lt;FontPlatformData::Attributes&gt;::decode):
    * Source/WebKit/Shared/WebCoreArgumentCoders.h:
    * Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
    (IPC::ArgumentCoder&lt;CFStringRef&gt;::encode):
    (IPC::ArgumentCoder&lt;RetainPtr&lt;CFStringRef&gt;&gt;::decode):

    Add FontPlatformData::Attributes for the POD serializable subset.

    Similar to adding Font::Attributes, this adds FontPlatformData::Attributes for the serializable data,
    that can be combined with an (optional) FontCustomPlatformData in order to create a final object.

    Unfortunately this can&apos;t actually be used as the main storage within FontPlatformData, as we need
    to do some data conversion before serialization (and the equivalent conversion on deserialization
    can&apos;t be done without the FontCustomPlatformData).

    * Source/WebCore/platform/graphics/FontPlatformData.h:
    (WebCore::FontPlatformData::Attributes::Attributes):
    * Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
    (WebCore::findFontDescriptor):
    (WebCore::createCTFont):
    (WebCore::FontPlatformData::create):
    (WebCore::FontPlatformData::attributes const):
    * Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
    (IPC::ArgumentCoder&lt;WebCore::FontPlatformData::Attributes&gt;::encodePlatformData):
    (IPC::ArgumentCoder&lt;WebCore::FontPlatformData::Attributes&gt;::decodePlatformData):
    * Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
    (IPC::ArgumentCoder&lt;WebCore::FontPlatformData::Attributes&gt;::encode):
    (IPC::ArgumentCoder&lt;FontPlatformData::Attributes&gt;::decode):
    * Source/WebKit/Shared/WebCoreArgumentCoders.h:
    * Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
    (IPC::ArgumentCoder&lt;CFStringRef&gt;::encode):
    (IPC::ArgumentCoder&lt;RetainPtr&lt;CFStringRef&gt;&gt;::decode):

    Make RemoteRenderingData::cacheFont reconstruct a Font using Attributes, and a reference to FontCustomPlatformData.

    Uses the new infrastructure to serialize a font using the Attributes of the Font and FontPlatformData, plus
    optionally a reference to an existing cached FontCustomPlatformData.

    We can then reconstruct a Font instance in the GPU process, using the shared FontCustomPlatformData and avoid
    re-serializing and parsing the data.

    * Source/WebCore/platform/graphics/Font.cpp:
    (WebCore::Font::description const):
    * Source/WebCore/platform/graphics/Font.h:
    * Source/WebCore/platform/graphics/FontPlatformData.h:
    * Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
    * Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
    (WebKit::RemoteRenderingBackend::cacheFont):
    * Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
    * Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
    * Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
    (WebKit::RemoteRenderingBackendProxy::cacheFont):
    * Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
    * Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
    (WebKit::RemoteResourceCacheProxy::recordFontUse):
</pre>

litherum commented on Source/WebCore/platform/graphics/FontCustomPlatformData.h:
```diff
@@ -74,17 +76,21 @@ struct FontCustomPlatformData {
 #elif USE(CORE_TEXT)
     RetainPtr<CTFontDescriptorRef> fontDescriptor;
-    FontPlatformData::CreationData creationData;
+    FontPlatformData::CreationData m_creationData;
```
FontCustomPlatformData is a struct, so its members shouldn't have the `m_` prefix.

litherum commented on Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
```diff
@@ -619,6 +619,50 @@
+    }
+    encoder << customPlatformData.creationData()->fontFaceData->size();
+    encoder << WTFMove(handle);
+    encoder << customPlatformData.creationData()->itemInCollection;
```
I'm kind of confused. `FontCustomPlatformData` includes a fontDescriptor member, but it isn't serialized, and it isn't recreated during deserialization either. So this kind of breaks the transparency that IPC is supposed to have.

mattwoodrow replied:
The fontDescriptor is a derived object, and a new one gets constructed in `createFontCustomPlatformData` when we decode.

litherum commented on Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
```diff
@@ -48,6 +48,59 @@ FontPlatformData::FontPlatformData(RetainPtr<CTFontRef>&& font, float size, bool
 #endif
 }
 
+static RetainPtr<CTFontDescriptorRef> findFontDescriptor(CFStringRef referenceURL, CFStringRef postScriptName)
```
Why duplicate these from WK2 to WebCore? This isn't explained in the commit message.

mattwoodrow replied:
We're now reconstructing the final font ourselves by asking the font code, rather than having the IPC layer do it.

litherum commented on Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
```diff
@@ -650,6 +650,12 @@ std::optional<RetainPtr<CFNumberRef>> ArgumentCoder<RetainPtr<CFNumberRef>>::dec
 template<typename Encoder>
 void ArgumentCoder<CFStringRef>::encode(Encoder& encoder, CFStringRef string)
 {
+    if (!string) {
+        encoder << true;
+        return;
+    }
+    encoder << false;
```
This seems unrelated

mattwoodrow replied:
Without this, serialising a `RetainPtr<CFString>` with `nullptr` value crashes, and that's a case we now hit. This just copies what the `CFDictionary` impl above does to handle it.

litherum commented on Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
```diff
@@ -671,6 +677,14 @@ template void ArgumentCoder<CFStringRef>::encode<StreamConnectionEncoder>(Stream
 
 std::optional<RetainPtr<CFStringRef>> ArgumentCoder<RetainPtr<CFStringRef>>::decode(Decoder& decoder)
 {
+    std::optional<bool> isNull;
+    decoder >> isNull;
+    if (!isNull)
+        return std::nullopt;
+
+    if (*isNull)
+        return { { nullptr } };
```
Ditto.

litherum commented on Source/WebCore/platform/graphics/FontPlatformData.h:
```diff
@@ -123,6 +124,36 @@ class FontPlatformData {
     FontPlatformData(cairo_font_face_t*, RefPtr<FcPattern>&&, float size, bool fixedWidth, bool syntheticBold, bool syntheticOblique, FontOrientation);
 #endif
 
+    class Attributes {
```
I don't really understand why this is better.

litherum replied:
Oh, I see, this is so the last patch can do

```
CacheFont(WebCore::Font::Attributes data, WebCore::FontPlatformData::Attributes platformData, std::optional<WebCore::RenderingResourceIdentifier> renderingResourceIdentifier) NotStreamEncodable
```

litherum commented on Source/WebCore/platform/graphics/Font.h:
```diff
@@ -215,6 +215,18 @@ class Font : public RefCounted<Font> {
     void setIsUsedInSystemFallbackFontCache() { m_isUsedInSystemFallbackFontCache = true; }
     bool isUsedInSystemFallbackFontCache() const { return m_isUsedInSystemFallbackFontCache; }
 
+    class Attributes {
```
I don't really understand why this is better.

litherum replied:
Oh, I see, this is so the last patch can do

```
CacheFont(WebCore::Font::Attributes data, WebCore::FontPlatformData::Attributes platformData, std::optional<WebCore::RenderingResourceIdentifier> renderingResourceIdentifier) NotStreamEncodable
```

litherum commented on Source/WebCore/platform/graphics/Font.cpp:
```diff
@@ -520,15 +520,13 @@ bool Font::isProbablyOnlyUsedToRenderIcons() const
 
 #endif
 
-#if !LOG_DISABLED
```
Why?

litherum commented on Source/WebCore/platform/graphics/Font.h:
```diff
@@ -184,9 +184,7 @@ class Font : public RefCounted<Font> {
     Visibility visibility() const { return m_attributes.visibility; }
     bool allowsAntialiasing() const { return m_allowsAntialiasing; }
 
-#if !LOG_DISABLED
```
Ditto.

litherum commented on Source/WebCore/platform/graphics/FontPlatformData.h:
```diff
@@ -242,7 +242,7 @@ class FontPlatformData {
     RefPtr<SharedBuffer> platformOpenTypeTable(uint32_t table) const;
 
-    String description() const;
+    WEBCORE_EXPORT String description() const;
```
This doesn't appear to be called from WK2

litherum commented on Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
```diff
@@ -204,8 +204,6 @@ RefPtr<SharedBuffer> FontPlatformData::openTypeTable(uint32_t table) const
     return platformOpenTypeTable(table);
 }
 
-#if !LOG_DISABLED
```
Same thing

litherum commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -345,12 +345,19 @@ void RemoteRenderingBackend::cacheNativeImageWithQualifiedIdentifier(const Share
+void RemoteRenderingBackend::cacheFont(const WebCore::Font::Attributes& fontAttributes, WebCore::FontPlatformData::Attributes platformData, std::optional<WebCore::RenderingResourceIdentifier> renderingResourceIdentifier)
 {
-    // Immediately turn the RenderingResourceIdentifier (which is error-prone) to a QualifiedRenderingResourceIdentifier,
-    // and use a helper function to make sure that don't accidentally use the RenderingResourceIdentifier (because the helper function can't see it).
```
I think you should keep this comment/behavior.

litherum commented on Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
```diff
@@ -137,10 +137,14 @@ void RemoteResourceCacheProxy::recordNativeImageUse(NativeImage& image)
 
 void RemoteResourceCacheProxy::recordFontUse(Font& font)
 {
+    if (font.platformData().customPlatformData())
+        recordFontCustomPlatformDataUse(*font.platformData().customPlatformData());
```
I guess this is fine. The crux of this set of patches is that you're splitting a font into 2 objects and serializing them separately, with different lifetimes, which naively seems unintuitive and overengineered, but if the memory savings make it worth it, then I guess it's fine. The repurposing of `FontCustomPlatformData` into a retained object that gets IPC'ed around, rather than just an implementation detail of how fonts get created internally, is a bit scary, but I can't think of a better way to do it without a major architecture change.

mattwoodrow committed:
Make RemoteRenderingData::cacheFont reconstruct a Font using Attributes, and a reference to FontCustomPlatformData.
https://bugs.webkit.org/show_bug.cgi?id=254490
<rdar://106965215>

Reviewed by Myles C. Maxfield.

Make FontCustomPlatformData refcounted, and owned by the FontPlatformData instead of CreationData

We currently only have the CreationData accessible from Font, but when serializing a web font,
we need the FontCustomPlatformData in order to share the parsed font data.

This just makes the FontCustomPlatformData the shared object on FontPlatformData, and we can
still access the CreationData through it.

* Source/WebCore/css/CSSFontFaceSource.h:
* Source/WebCore/loader/cache/CachedFont.cpp:
(WebCore::CachedFont::createCustomFontData):
* Source/WebCore/loader/cache/CachedFont.h:
* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
(WebCore::FontCustomPlatformData::FontCustomPlatformData):
* Source/WebCore/platform/graphics/FontPlatformData.cpp:
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::FontPlatformData::creationData const):
* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformData::customPlatformData const):
(WebCore::FontPlatformData::creationData const): Deleted.
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::FontCache::systemFallbackForCharacters):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::createDerivativeFont):
(WebCore::Font::createFontWithoutSynthesizableFeatures const):
(WebCore::Font::platformCreateScaledFont const):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::FontPlatformData::FontPlatformData):
* Source/WebCore/platform/graphics/mac/FontCustomPlatformDataMac.cpp:
(WebCore::FontCustomPlatformData::fontPlatformData):
(WebCore::createFontCustomPlatformData):
(WebCore::FontCustomPlatformData::creationData const):
* Source/WebCore/workers/WorkerFontLoadRequest.h:
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder<WebCore::Font>::decodePlatformData):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:

Move POD serializable subset of Font into Attributes object

In order to serialize a Font where the FontCustomPlatformData is already cached, we want to send
the identifier of the data, plus the serialized remainder of the font.

This creates an Attributes inner class for Font, containing the data needed to serialize the font,
without the platform data.

* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::Font):
(WebCore::Font::renderingResourceIdentifier const):
* Source/WebCore/platform/graphics/Font.h:
(WebCore::Font::isTextOrientationFallback const):
(WebCore::Font::origin const):
(WebCore::Font::isInterstitial const):
(WebCore::Font::visibility const):
(WebCore::Font::attributes const):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<WebCore::Font>::encode):
(IPC::ArgumentCoder<Font>::decode):
(IPC::ArgumentCoder<WebCore::Font::Attributes>::encode):
(IPC::ArgumentCoder<Font::Attributes>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:

Add serialization and RemoteResourceCache support for FontCustomPlatformData

This lets us cache FontCustomPlatformData objects independently of Font, so that we
can share them between Font instances in the GPUP.

* Source/WebKit/GPUProcess/graphics/QualifiedResourceHeap.h:
(WebKit::QualifiedResourceHeap::add):
(WebKit::QualifiedResourceHeap::getFontCustomPlatformData const):
(WebKit::QualifiedResourceHeap::removeFontCustomPlatformData):
(WebKit::QualifiedResourceHeap::releaseAllResources):
(WebKit::QualifiedResourceHeap::checkInvariants const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cacheFontCustomPlatformData):
(WebKit::RemoteRenderingBackend::cacheFontCustomPlatformDataWithQualifiedIdentifier):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cacheFontCustomPlatformData):
(WebKit::RemoteResourceCache::cachedFontCustomPlatformData const):
(WebKit::RemoteResourceCache::releaseRenderingResource):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder<WebCore::FontCustomPlatformData>::encodePlatformData):
(IPC::ArgumentCoder<WebCore::FontCustomPlatformData>::decodePlatformData):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<WebCore::FontCustomPlatformData>::encode):
(IPC::ArgumentCoder<FontCustomPlatformData>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheFontCustomPlatformData):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordFontCustomPlatformDataUse):
(WebKit::RemoteResourceCacheProxy::prepareForNextRenderingUpdate):
(WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdateForFonts):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

Add serialization and RemoteResourceCache support for FontCustomPlatformData

This lets us cache FontCustomPlatformData objects independently of Font, so that we
can share them between Font instances in the GPUP.

* Source/WebKit/GPUProcess/graphics/QualifiedResourceHeap.h:
(WebKit::QualifiedResourceHeap::add):
(WebKit::QualifiedResourceHeap::getFontCustomPlatformData const):
(WebKit::QualifiedResourceHeap::removeFontCustomPlatformData):
(WebKit::QualifiedResourceHeap::releaseAllResources):
(WebKit::QualifiedResourceHeap::checkInvariants const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cacheFontCustomPlatformData):
(WebKit::RemoteRenderingBackend::cacheFontCustomPlatformDataWithQualifiedIdentifier):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cacheFontCustomPlatformData):
(WebKit::RemoteResourceCache::cachedFontCustomPlatformData const):
(WebKit::RemoteResourceCache::releaseRenderingResource):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder<WebCore::FontCustomPlatformData>::encodePlatformData):
(IPC::ArgumentCoder<WebCore::FontCustomPlatformData>::decodePlatformData):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<WebCore::FontCustomPlatformData>::encode):
(IPC::ArgumentCoder<FontCustomPlatformData>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheFontCustomPlatformData):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordFontCustomPlatformDataUse):
(WebKit::RemoteResourceCacheProxy::prepareForNextRenderingUpdate):
(WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdateForFonts):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

Add FontPlatformData::Attributes for the POD serializable subset.

Similar to adding Font::Attributes, this adds FontPlatformData::Attributes for the serializable data,
that can be combined with an (optional) FontCustomPlatformData in order to create a final object.

Unfortunately this can't actually be used as the main storage within FontPlatformData, as we need
to do some data conversion before serialization (and the equivalent conversion on deserialization
can't be done without the FontCustomPlatformData).

* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformData::Attributes::Attributes):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::findFontDescriptor):
(WebCore::createCTFont):
(WebCore::FontPlatformData::create):
(WebCore::FontPlatformData::attributes const):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder<WebCore::FontPlatformData::Attributes>::encodePlatformData):
(IPC::ArgumentCoder<WebCore::FontPlatformData::Attributes>::decodePlatformData):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<WebCore::FontPlatformData::Attributes>::encode):
(IPC::ArgumentCoder<FontPlatformData::Attributes>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::ArgumentCoder<CFStringRef>::encode):
(IPC::ArgumentCoder<RetainPtr<CFStringRef>>::decode):

Add FontPlatformData::Attributes for the POD serializable subset.

Similar to adding Font::Attributes, this adds FontPlatformData::Attributes for the serializable data,
that can be combined with an (optional) FontCustomPlatformData in order to create a final object.

Unfortunately this can't actually be used as the main storage within FontPlatformData, as we need
to do some data conversion before serialization (and the equivalent conversion on deserialization
can't be done without the FontCustomPlatformData).

* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformData::Attributes::Attributes):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::findFontDescriptor):
(WebCore::createCTFont):
(WebCore::FontPlatformData::create):
(WebCore::FontPlatformData::attributes const):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder<WebCore::FontPlatformData::Attributes>::encodePlatformData):
(IPC::ArgumentCoder<WebCore::FontPlatformData::Attributes>::decodePlatformData):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<WebCore::FontPlatformData::Attributes>::encode):
(IPC::ArgumentCoder<FontPlatformData::Attributes>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::ArgumentCoder<CFStringRef>::encode):
(IPC::ArgumentCoder<RetainPtr<CFStringRef>>::decode):

Make RemoteRenderingData::cacheFont reconstruct a Font using Attributes, and a reference to FontCustomPlatformData.

Uses the new infrastructure to serialize a font using the Attributes of the Font and FontPlatformData, plus
optionally a reference to an existing cached FontCustomPlatformData.

We can then reconstruct a Font instance in the GPU process, using the shared FontCustomPlatformData and avoid
re-serializing and parsing the data.

* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::description const):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cacheFont):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheFont):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordFontUse):

Canonical link: https://commits.webkit.org/262263@main
